wazo_agid/handlers/outgoing_callerid_formatter.py

```diff
@@ -0,0 +1,87 @@
+# Copyright 2024 The Wazo Authors  (see the AUTHORS file)
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+import re
+
+import phonenumbers
+
+from wazo_agid import objects
+from wazo_agid.handlers import handler
+
+VALID_PHONE_NUMBER_RE = re.compile(r'^\+?\d{3,15}$')
+CALLER_ID_ALL_REGEX = re.compile(r'^"(.*)" <(\+?\d{3,15})>$')
+
+
+def _remove_none_numeric_char(raw: str) -> str:
+    return ''.join(c for c in raw if c.isdigit())
+
+
+class CallerIDFormatter(handler.Handler):
+    def execute(self) -> None:
+        self.set_caller_id()
+
+    def set_caller_id(self) -> None:
+        selected_cid = self._agi.get_variable('WAZO_SELECTED_CALLER_ID_TO_FORMAT')
+        cid_format = self._agi.get_variable('TRUNK_OUTGOING_CALLER_ID_FORMAT')
+        tenant_country = self._agi.get_variable('WAZO_TENANT_COUNTRY')
+
+        if not selected_cid:
+            return
+
+        if not cid_format:
+            return
+
+        matches = CALLER_ID_ALL_REGEX.match(selected_cid)
+        if matches:
+            cid_name = matches.group(1)
+            cid_number = matches.group(2)
+        else:
+            cid_name = ''
+            cid_number = selected_cid
+
+        try:
+            cid_number = phonenumbers.parse(cid_number, tenant_country)
+        except phonenumbers.phonenumberutil.NumberParseException:
+            self._set_raw_number(cid_name, cid_number)
+        else:
+            self._set_formated_number(cid_name, cid_number, cid_format)
+
+    def _set_raw_number(self, name: str, number: str) -> None:
+        matches = VALID_PHONE_NUMBER_RE.match(number)
```

Comment:
Why do we do this second level of validation here, after already trying to parse the number with `phonenumbers.parse`?

Reply:
Ah right, avoid injection attack